url49526af437_Maria_G_CS325_tgz-pJ8-assignments_farmer_FarmerImplJ8/src/minidraw/standard/ImageManager.java

```diff
@@ -0,0 +1,154 @@
+    As the ImageManager is a singleton class you access it using 
+    <tt>ImageManager.getSingleton()</tt>. The MinimalApplication
+    creates the ImageManager automatically.
+
+   This source code is from the book 
+     "Flexible, Reliable Software:
+       Using Patterns and Agile Development"
+     published 2010 by CRC Press.
+   Author: 
+     Henrik B Christensen 
+     Computer Science Department
+     Aarhus University
+   
+   This source code is provided WITHOUT ANY WARRANTY either 
+   expressed or implied. You may study, use, modify, and 
+   distribute it for non-commercial purposes. For any 
+   commercial use, see http://www.baerbak.com/
+*/
+    
+public class ImageManager {
+
+  private static final String RESOURCE_PATH      = "/resource/";
+  private static Component aComponent;
+
+  private Map<String,Image> name2Image;
+  
+  public static ImageManager singleton;
+
+  ImageManager(Component c) {
+    aComponent = c;
+    registerAllImages();
+    singleton = this;
+  }
+
+  public static ImageManager getSingleton() {
+    if ( singleton == null ) {
+      throw new RuntimeException("getSingleton() invoked before ImageManger "
+                                 + "has been instantiated." );
+    }
+    return singleton;
+  }
+
+  /** load all gif files in resource folder into the image manager. */
+  private void registerAllImages() {
+    name2Image = new Hashtable<String,Image>();
+
+    MediaTracker tracker = new MediaTracker(aComponent);
+
+    try {
+      java.net.URL _url = getClass().getResource(RESOURCE_PATH);
+      if ( _url == null ) {
+        throw new RuntimeException("ImageManager: URL/folder '"+
+                                   RESOURCE_PATH+
+                                   "' does not exist.");
+      }
+      // Convert URL to a File (from Kohsuke Kawaguchi's Blog)
+      File dir;
+      try {
+        dir = new File(_url.toURI());
+      } catch(URISyntaxException e) {
+        dir = new File(_url.getPath());
+      }
+      // create a filter to find .GIF files.
+      FilenameFilter filter = new FilenameFilter() {
+        public boolean accept(File dir, String name) {
+            return name.endsWith(".gif") || name.endsWith(".GIF");
+        }
+      };
+      // Retrieve array of gif file names in resource folder
+      String[] gifFileArray = dir.list(filter);
+      for ( String s : gifFileArray ) {
+        // Load and register the image
+        Image img = _registerIn(s);
+        if ( img == null ) {
+          throw new 
+            RuntimeException("ImageManager: Did not find image named "+
+                             s );
+        }
+        tracker.addImage( img, 7 /* not used so why not 7 */ );
+        String filenameNoExtension = s.substring(0, s.length()-4);
+        name2Image.put( filenameNoExtension, img );        
+      }
+      // block until all images are loaded
+      try {
+        tracker.waitForAll();
+      } catch (Exception e) {}
+    } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)